gnomad/utils/transcript_annotation.py

```diff
@@ -205,3 +214,146 @@ def tissue_expression_ht_to_array(
+    :param tx_ht: Table of transcript expression information.
+    :param tissues_to_filter: Optional list of tissues to exclude from the output.
+    :param vep_root: Name used for root VEP annotation. Default is 'vep'.
+    :param vep_annotation: Name of annotation in 'vep' annotation,
+        one of the processed consequences: ["transcript_consequences",
+        "worst_csq_by_gene", "worst_csq_for_variant",
+        "worst_csq_by_gene_canonical", "worst_csq_for_variant_canonical"].
+        For example, if you want to annotate each variant with the worst
+        consequence in each gene it falls on and the transcript expression,
+        you would use "worst_csq_by_gene". Default is "transcript_consequences".
+    :return: Input Table with transcript expression information annotated.
+    """
+    # Filter to tissues of interest.
+    tx_ht = filter_expression_ht_by_tissues(tx_ht, tissues_to_filter=tissues_to_filter)
+    tissues = list(tx_ht.row_value)
+
+    # Calculate the mean expression proportion across all tissues.
+    tx_ht = tx_ht.annotate(
+        exp_prop_mean=hl.mean([tx_ht[t].expression_proportion for t in tissues])
+    )
+
+    # Explode the processed transcript consequences to be able to key by
+    # transcript ID.
+    ht = explode_by_vep_annotation(ht, vep_annotation=vep_annotation, vep_root=vep_root)
+    ht = ht.transmute(
+        **ht[vep_annotation],
+        **tx_ht[ht[vep_annotation].transcript_id, ht[vep_annotation].gene_id],
+    )
+    ht = ht.annotate_globals(tissues=tissues)
+
+    return ht
+
+
+def tx_aggregate_variants(
+    ht: hl.Table,
+    additional_grouping: bool = False,
+    additional_group_by: Optional[Union[Tuple[str], List[str]]] = (
+        "alleles",
+        "gene_symbol",
+        "most_severe_consequence",
+        "lof",
+        "lof_flags",
+    ),
+) -> hl.Table:
+    """
+    Aggregate transcript-based expression values or expression proportion from GTEx.
+
+    :param ht: Table of variants annotated with transcript expression information.
+    :param additional_grouping: Whether to group by additional fields before sum
+        aggregation. Default is False.
+    :param additional_group_by: Optional list of additional fields to group by before
+        sum aggregation.
+    :return: Table of variants with transcript expression information aggregated.
+    """
+    tissues = hl.eval(ht.tissues)
+
+    grouping = (
+        ["locus", "gene_id"] + list(additional_group_by)
+        if additional_grouping
+        else ["locus", "gene_id"]
+    )
+
+    # Aggregate the transcript expression information by locus, gene_id and
+    # annotations in additional_group_by.
+    ht = ht.group_by(*grouping).aggregate(
+        exp_prop_mean=hl.agg.sum(ht.exp_prop_mean),
+        **{t: hl.struct(**{a: hl.agg.sum(ht[t][a]) for a in ht[t]}) for t in tissues},
+    )
+
+    ht = ht.key_by(ht.locus, ht.alleles) if additional_grouping else ht.key_by(ht.locus)
+
+    return ht
```

Comment:
I wonder if it might be helpful to add a function that combines all of these too? Something like:

```
def some_name(
    ht: hl.Table,
    tx_ht: hl.Table,
    tissues_to_filter: Optional[List[str]] = None,
    vep_root: str = "vep",
    vep_annotation: str = "transcript_consequences",
    filter_to_csqs=CSQ_CODING,
    additional_group_by: Optional[Union[Tuple[str], List[str]]] = (
        "alleles",
        "gene_symbol",
        "most_severe_consequence",
        "lof",
        "lof_flags",
    ),
    **kwargs,
) -> hl.Table:
    tx_ht = tx_annotate_variants(
        preprocess_variants_for_tx(
            ht, vep_root=vep_root, filter_to_csqs=filter_to_csqs, **kwargs
        ),
        tx_ht,
        tissues_to_filter=tissues_to_filter,
        vep_root=vep_root,
        vep_annotation=vep_annotation,
    )
    tx_ht = tx_aggregate_variants(tx_ht, additional_group_by=additional_group_by)

    tx_ht = tx_ht.collect_by_key("tx_annotation")
    ht = ht.annotate(**tx_ht[ht.key])

    return ht
```

Reply:
I'm not sure if we should make this function here or just in gnomad_qc/v*/annotations/pext.py.

Reply:
I like it here, it is general enough to be in gnomad_methods. I also completely agree with you on not annotating the original Table.